File: 02-dataclasses/filesystem.py
from dataclasses import dataclass, field
from datetime import datetime


@dataclass
class File:
    name: str
    last_modified: datetime
    readonly: bool = False


@dataclass
class Directory:
    name: str
    files: list[File] = field(default_factory=list)
    subdirectories: list['Directory'] = field(default_factory=list)


project = Directory('PythonFromAltUniverses',
    subdirectories=[
        Directory('02-dataclasses',
            files=[
                File('filesystem.py', datetime(2022, 3, 15, 18, 9, 40)),
                File('address.py', datetime(2022, 3, 15, 17, 50, 0), readonly=True)
            ]
        )
    ]
)

print(project.subdirectories[0].files[0].name)
print(project)

File: 5-dataclasses/filesystem.cpp
#include <iostream>
#include <string>
#include <vector>


struct File {
    std::string name;
    double last_modified;
    bool readonly = false;
};


struct Directory {
    std::string name;
    std::vector<File> files;
    std::vector<Directory> subdirectories;
};


int main() {
    Directory const project{"PythonFromAltUniverses",
        {},
        {
            Directory{"5-dataclasses",
                {
                    File{"filesystem.py", 1647339391.0},
                    File{"address.py", 1647339402.0, true}
                }
            }
        }
    };

    std::cout << project.subdirectories[0].files[0].name << std::endl;
}
